pinot-connectors/pinot-spark-3-connector/src/main/scala/org/apache/pinot/connector/spark/v3/datasource/PinotDataWriter.scala

```diff
@@ -0,0 +1,246 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.connector.spark.v3.datasource
+
+import org.apache.commons.io.FileUtils
+import org.apache.pinot.common.utils.TarGzCompressionUtils
+import org.apache.pinot.connector.spark.common.PinotDataSourceWriteOptions
+import org.apache.spark.sql.connector.write.{DataWriter, WriterCommitMessage}
+import org.apache.pinot.segment.local.segment.creator.impl.SegmentIndexCreationDriverImpl
+import org.apache.pinot.segment.spi.creator.SegmentGeneratorConfig
+import org.apache.pinot.spi.config.table.{IndexingConfig, SegmentsValidationAndRetentionConfig, TableConfig, TableCustomConfig, TenantConfig}
+import org.apache.pinot.spi.data.readers.GenericRow
+import org.apache.pinot.spi.data.Schema
+import org.apache.pinot.spi.ingestion.batch.spec.Constants
+import org.apache.pinot.spi.utils.DataSizeUtils
+import org.apache.spark.sql.catalyst
+import org.apache.spark.sql.types.StructType
+import org.slf4j.{Logger, LoggerFactory}
+
+import java.io.File
+import java.nio.file.Files
+import java.util.regex.Pattern
+
+class PinotDataWriter[InternalRow](
+                                    partitionId: Int,
+                                    taskId: Long,
+                                    writeOptions: PinotDataSourceWriteOptions,
+                                    writeSchema: StructType,
+                                    pinotSchema: Schema)
+  extends DataWriter[org.apache.spark.sql.catalyst.InternalRow] with AutoCloseable {
+  private val logger: Logger = LoggerFactory.getLogger(classOf[PinotDataWriter[InternalRow]])
+  logger.info("PinotDataWriter created with writeOptions: {}, partitionId: {}, taskId: {}",
+    (writeOptions, partitionId, taskId))
+
+  val tableName: String = writeOptions.tableName
+  val savePath: String = writeOptions.savePath
+  val bufferedRecordReader: PinotBufferedRecordReader = new PinotBufferedRecordReader()
+
+  override def write(record: catalyst.InternalRow): Unit = {
+    bufferedRecordReader.write(internalRowToGenericRow(record))
```

Comment:
I think min/max timestamp is not sufficient. The main use-case I wanted us to support was "string partitions".

This can be a common requirement where a Pinot table is used by another platform which allows users to define custom datasets with a pre-defined schema (e.g. our segmentation use-case I shared in the last year's Uber Meetup).

> a serialized class

I didn't get this, can we expose something like this?

```
airlineStats.write.format("pinot") \
      .mode("append") \
      .option("table", "airlineStats") \
      .option("segmentNameGenerator", "com.mycompany.foo.MySegmentNameGenerator")
      .save("/my/segments/path")
```

Reply:
I see. I think it should be doable but seems like a non-trivial feature and requires some thinking.
What do you say if I start with this template based approach and come back to this once we have the basics working?

Reply:
Yeah that sounds reasonable to me